jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -853,110 +853,110 @@ private static final Object resolveConstantDynamic(long j9class, String name, St
-						cpEntry = (char) cpValue;
-					} else {
-						cpEntry = cpValue;
-					}
-					break;
+		/* Check if we need to treat the last parameter specially when handling primitives.
+		 * The type of the varargs array will determine how primitive ints from the constantpool
+		 * get boxed: {Boolean, Byte, Short, Character or Integer}.
+		 */
+		boolean treatLastArgAsVarargs = bsm.isVarargsCollector();
+		Class<?> varargsComponentType = bsm.type.lastParameterType().getComponentType();
+		int bsmTypeArgCount = bsm.type.parameterCount();
+
+		for (int i = 0; i < bsmArgCount; i++) {
+			int staticArgIndex = BSM_OPTIONAL_ARGUMENTS_START_INDEX + i;
+			short index = UNSAFE.getShort(bsmArgs + (i * BSM_ARGUMENT_SIZE));
+			int cpType = getCPTypeAt(internalRamClass, index);
+			Object cpEntry = null;
+			switch (cpType) {
+			case 1:
+				cpEntry = cp.getClassAt(index);
+				if (cpEntry == null) {
+					throw throwNoClassDefFoundError(classObject, index);
 				}
-				case 4:
-					cpEntry = cp.getFloatAt(index);
-					break;
-				case 5:
-					cpEntry = cp.getLongAt(index);
-					break;
-				case 6:
-					cpEntry = cp.getDoubleAt(index);
-					break;
-				case 13:
-					cpEntry = getCPMethodTypeAt(internalRamClass, index);
-					break;
-				case 14:
-					cpEntry = getCPMethodHandleAt(internalRamClass, index);
-					break;
-				case 17:
-					cpEntry = getCPConstantDynamicAt(internalRamClass, index);
-					break;
-				default:
-					// Do nothing. The null check below will throw the appropriate exception.
+				break;
+			case 2:
+				cpEntry = cp.getStringAt(index);
+				break;
+			case 3: {
+				int cpValue = cp.getIntAt(index);
+				Class<?> argClass;
+				if (treatLastArgAsVarargs && (staticArgIndex >= (bsmTypeArgCount - 1))) {
+					argClass = varargsComponentType;
+				} else {
+					argClass = bsm.type.parameterType(staticArgIndex);
 				}
-				
-				cpEntry.getClass();	// Implicit NPE
-				staticArgs[staticArgIndex] = cpEntry;
+				if (argClass == Short.TYPE) {
+					cpEntry = (short) cpValue;
+				} else if (argClass == Boolean.TYPE) {
+					cpEntry = cpValue == 0 ? Boolean.FALSE : Boolean.TRUE;
+				} else if (argClass == Byte.TYPE) {
+					cpEntry = (byte) cpValue;
+				} else if (argClass == Character.TYPE) {
+					cpEntry = (char) cpValue;
+				} else {
+					cpEntry = cpValue;
+				}
+				break;
+			}
+			case 4:
+				cpEntry = cp.getFloatAt(index);
+				break;
+			case 5:
+				cpEntry = cp.getLongAt(index);
+				break;
+			case 6:
+				cpEntry = cp.getDoubleAt(index);
+				break;
+			case 13:
+				cpEntry = getCPMethodTypeAt(internalRamClass, index);
+				break;
+			case 14:
+				cpEntry = getCPMethodHandleAt(internalRamClass, index);
+				break;
+			case 17:
+				cpEntry = getCPConstantDynamicAt(internalRamClass, index);
+				break;
+			default:
+				// Do nothing. The null check below will throw the appropriate exception.
 			}
 
+			cpEntry.getClass();	// Implicit NPE
+			staticArgs[staticArgIndex] = cpEntry;
+		}
+
+		try {
```

Comment:
Can you add a comment to this try block that references the appropriate part of the JVMS so that we can double check this behaviour in the future?  It makes it much easier to figure out than trying to discern the subtle difference from the code